src/t_list.c

```diff
@@ -842,47 +842,57 @@ robj *getStringObjectFromListPosition(int position) {
     }
 }
 
-void lmoveGenericCommand(client *c, int wherefrom, int whereto) {
-    robj *sobj, *value;
-    if ((sobj = lookupKeyWriteOrReply(c,c->argv[1],shared.null[c->resp]))
-        == NULL || checkType(c,sobj,OBJ_LIST)) return;
-
-    if (listTypeLength(sobj) == 0) {
-        /* This may only happen after loading very old RDB files. Recent
-         * versions of Redis delete keys of empty lists. */
-        addReplyNull(c);
-    } else {
-        robj *dobj = lookupKeyWrite(c->db,c->argv[2]);
-        robj *touchedkey = c->argv[1];
+void lmmoveGenericCommand(client *c, robj **src, long numsrc, robj *dst, int wherefrom, int whereto) {
+    for (int j = 0; j < numsrc; j++) {
+        robj *sobj = lookupKeyWrite(c->db,src[j]);
+
+        if (sobj == NULL) continue;
+        if (checkType(c,sobj,OBJ_LIST)) return;
+        if (listTypeLength(sobj) == 0) {
+            /* This may only happen after loading very old RDB files. Recent
+            * versions of Redis delete keys of empty lists. */
+            continue;
+        }
 
+        robj *dobj = lookupKeyWrite(c->db,dst);
         if (checkType(c,dobj,OBJ_LIST)) return;
-        value = listTypePop(sobj,wherefrom);
+
+        robj *value = listTypePop(sobj,wherefrom);
         serverAssert(value); /* assertion for valgrind (avoid NPD) */
-        lmoveHandlePush(c,c->argv[2],dobj,value,whereto);
+        lmoveHandlePush(c,dst,dobj,value,whereto);
 
         /* listTypePop returns an object with its refcount incremented */
         decrRefCount(value);
 
         /* Delete the source list when it is empty */
         notifyKeyspaceEvent(NOTIFY_LIST,
                             wherefrom == LIST_HEAD ? "lpop" : "rpop",
-                            touchedkey,
+                            src[j],
                             c->db->id);
         if (listTypeLength(sobj) == 0) {
-            dbDelete(c->db,touchedkey);
+            dbDelete(c->db,src[j]);
             notifyKeyspaceEvent(NOTIFY_GENERIC,"del",
-                                touchedkey,c->db->id);
+                                src[j],c->db->id);
         }
-        signalModifiedKey(c,c->db,touchedkey);
+        signalModifiedKey(c,c->db,src[j]);
         server.dirty++;
-        if (c->cmd->proc == blmoveCommand) {
+
+        if (c->cmd->proc == lmmoveCommand || c->cmd->proc == blmmoveCommand) {
+            int offset = c->cmd->proc == blmmoveCommand ? 1 : 0;
+            rewriteClientCommandVector(c,5,shared.lmove,
+                c->argv[3+j],c->argv[1],c->argv[c->argc-2-offset],c->argv[c->argc-1-offset]);
```

Comment:
We can't just find wherefrom/whereto by `c->argv[3+numsrc]`, becase `numsrc` might not be the real keycount if `lmmoveGenericCommand` was fast-forwarded by its blocking version. 
Either we use `argc` or calculate the offset `src - argv` with looks even more tricky. Instead of passing a shifted src pointer, we can also pass an additional offset. This would make calculating arguments easier